src/main/java/ua/team3/carsharingservice/service/impl/RentalServiceImpl.java

```diff
@@ -0,0 +1,114 @@
+package ua.team3.carsharingservice.service.impl;
+
+import jakarta.persistence.EntityNotFoundException;
+import jakarta.transaction.Transactional;
+import java.time.LocalDate;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import ua.team3.carsharingservice.dto.RentalDto;
+import ua.team3.carsharingservice.dto.RentalRequestDto;
+import ua.team3.carsharingservice.exception.NoCarsAvailableException;
+import ua.team3.carsharingservice.exception.NotValidRentalDateException;
+import ua.team3.carsharingservice.exception.NotValidReturnDateException;
+import ua.team3.carsharingservice.exception.RentalAlreadyReturnedException;
+import ua.team3.carsharingservice.mapper.RentalMapper;
+import ua.team3.carsharingservice.model.Car;
+import ua.team3.carsharingservice.model.Rental;
+import ua.team3.carsharingservice.model.User;
+import ua.team3.carsharingservice.repository.CarRepository;
+import ua.team3.carsharingservice.repository.RentalRepository;
+import ua.team3.carsharingservice.service.RentalService;
+
+@Service
+@RequiredArgsConstructor
+public class RentalServiceImpl implements RentalService {
+    private final RentalRepository rentalRepository;
+    private final CarRepository carRepository;
+    private final RentalMapper rentalMapper;
+
+    @Override
+    public List<RentalDto> getAll(User user) {
+        return rentalRepository.findByUserId(user.getId())
+                .stream()
+                .map(rentalMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public RentalDto getById(Long id, User user) {
+        Rental rental = rentalRepository.findByIdAndUserId(id, user.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find a rental by id: " + id
+                ));
+        return rentalMapper.toDto(rental);
+    }
+
+    @Override
+    @Transactional
+    public RentalDto create(RentalRequestDto dto, User user) {
+        boolean hasActiveRentals = rentalRepository.findByUserId(user.getId()).stream()
+                .anyMatch(rental -> rental.getActualReturnDate() == null
+                        && rental.getReturnDate().isBefore(LocalDate.now()));
+
+        if (hasActiveRentals) {
+            throw new NoCarsAvailableException(
+                    "The user has overdue car rentals"
```

Comment:
We could put here the email or username for more specificity

Reply:
I'm not sure if it is needed, as the user already should know their email

Reply:
Maybe it is better to send `Sorry, you are not allowed to rent a car because you have an overdue rental car` ?

Reply:
Good idea